### ISSUE TITLE
`jnp.where` / `lax.select` reject broadcast-compatible shardings under explicit sharding

### ISSUE THREAD

**Issue by Gattocrucco** [jax-ml/jax#37853]
### Description

Under explicit sharding (`AxisType.Explicit`), elementwise ufuncs like `jnp.add` happily combine arrays of the same shape with different but broadcast-compatible shardings (e.g., a replicated operand and a sharded operand) and propagate the more refined sharding to the output. `jnp.where` in its elementwise (3-arg) form, and `lax.select`, do not: they raise `ShardingTypeError` unless `cond` has the exact same sharding as `a` and `b` (or is a scalar). This is a common elementwise operation so it is annoying when writing sharded code in explicit mode.

### Minimal reproducer

```python
import os
os.environ['XLA_FLAGS'] = '--xla_force_host_platform_device_count=4'

import jax
import jax.numpy as jnp
from jax import lax
from jax.sharding import PartitionSpec, AxisType

mesh = jax.make_mesh((4,), ('x',), axis_types=(AxisType.Explicit,))
sharded = jax.NamedSharding(mesh, PartitionSpec('x'))
replicated = jax.NamedSharding(mesh, PartitionSpec())

with jax.set_mesh(mesh):
    a = jax.device_put(jnp.zeros(16), sharded)
    b = jax.device_put(jnp.ones(16), sharded)
    cond = jax.device_put(jnp.arange(16) > 7, replicated)

    # Elementwise ufuncs combine broadcast-compatible shardings just fine:
    assert jnp.add(cond, a).sharding == sharded
    assert jnp.maximum(cond, a).sharding == sharded

    # But jnp.where / lax.select reject the same combination:
    for f in (lambda: jnp.where(cond, a, b), lambda: lax.select(cond, a, b)):
        try:
            f()
        except Exception as e:
            assert type(e).__name__ == 'ShardingTypeError'
            assert 'select `which` must be scalar or have the same sharding' in str(e)
            print(e)
```

```
select `which` must be scalar or have the same sharding as cases, got `which` sharding NamedSharding(mesh=AbstractMesh('x': 4, axis_types=(Explicit,), device_kind=cpu, num_cores=None), spec=P(None,)) but case sharding NamedSharding(mesh=AbstractMesh('x': 4, axis_types=(Explicit,), device_kind=cpu, num_cores=None), spec=P('x',)).
select `which` must be scalar or have the same sharding as cases, got `which` sharding NamedSharding(mesh=AbstractMesh('x': 4, axis_types=(Explicit,), device_kind=cpu, num_cores=None), spec=P(None,)) but case sharding NamedSharding(mesh=AbstractMesh('x': 4, axis_types=(Explicit,), device_kind=cpu, num_cores=None), spec=P('x',)).
```

This check lives at https://github.com/jax-ml/jax/blob/b32e191e22eeb8c2475a8866d91b51e34ab4bfc8/jax/_src/lax/lax.py#L7940-L7954

### Expected behavior

Even if not binary, I expected that any elementwise operation would have well-defined explicit sharding semantics, equivalent to `elementwise_sharding(a, b, c) = elementwise_sharding(a, elementwise_sharding(b, c))`. Is there a reason that isn't the case, or is it a missing implementation detail?

### Related

- google/flax#5209 hit the same error from `lax.select` inside `nnx.Dropout`; it was worked around in Flax rather than fixed in jax.
- #24520 originally added the `select_n_p` sharding rule with the current strict check.

### Note

Written by Opus 4.7, edited by a human.

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.10.0
jaxlib: 0.10.0
numpy:  2.4.6
python: 3.14.0 (main, Oct 14 2025, 21:10:22) [Clang 20.1.4 ]
device info: cpu-1, 1 local devices"
process_count: 1
platform: uname_result(system='Darwin', release='25.4.0', version='Darwin Kernel Version 25.4.0: Thu Mar 19 19:30:44 PDT 2026; root:xnu-12377.101.15~1/RELEASE_ARM64_T6000', machine='arm64')
```

**Comment by yashk2810** [comment]
Thanks for the issue!

I don't think `select` should allow broadcast compatible shardings because select's shape rules also don't allow broadcast compatible shapes!

But even if shapes were allowed to be broadcast compatible (which they are not), shardings can be more stricter as it as implications on backward pass comms.

> elementwise ufuncs like jnp.add happily combine arrays of the same shape with different but broadcast-compatible shardings

This is a good point! But maybe we shouldn't have allowed this because doing `P() + P('x') -> P('x')` means that on operand 0, there will be an AG on the bwd pass. These implicit comms are bad and shouldn't happen. Undoing the elementwise broadcast decision will require fixing a lot of places but the point still stands.

**Comment by Gattocrucco** [comment]
Ok, that's interesting, so these rules may actually go away instead of expand. Closing the issue then.
